sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs

```diff
@@ -0,0 +1,147 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2018 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Protobuf.Ucfg;
+using UcfgLocation = SonarAnalyzer.Protobuf.Ucfg.Location;
+
+namespace SonarAnalyzer.ControlFlowGraph.CSharp
+{
+    public class UcfgObjectFactory
+    {
+        /// <summary>
+        /// The string constant representation in the Sonar Security engine (Java part). When
+        /// an instruction receives or returns a type that is not string we use this instead
+        /// of a variable.
+        /// </summary>
+        private static readonly Expression ConstantExpression = new Expression
+        {
+            Const = new Constant { Value = "\"\"" }
+        };
+
+        private readonly Dictionary<SyntaxNode, Expression> nodeExpressionMap = new Dictionary<SyntaxNode, Expression>();
+
+        private int tempVariablesCounter;
+
+        public UCFG CreateUcfg(SyntaxNode syntaxNode, UcfgMethod method) =>
+            new UCFG
+            {
+                Location = GetLocation(syntaxNode),
+                MethodId = method.ToString(),
+            };
+
+        public BasicBlock CreateUcfgBlock(string id) =>
+            new BasicBlock { Id = id };
+
+        public Instruction CreateMethodCall(SyntaxNode syntaxNode, UcfgMethod method, params SyntaxNode[] arguments) =>
```

Comment:
All of the methods returning `Instruction` should belong to the `UcfgInstructionBuilder`.

Reply:
UcfgInstructionFactory is taking care of calling the correct method of the UcfgObjectFactory (which instantiates the Instruction class). The separation is nice imo because one class works with Syntax, other with UCFG objects.